router/pkg/routerplugin/grpc_plugin_client.go

```diff
@@ -16,27 +18,114 @@ type GRPCPluginClient struct {
 
 	pc *plugin.Client
 	cc grpc.ClientConnInterface
+
+	config GRPCPluginClientConfig
+
+	mu sync.RWMutex
+}
+
+type GRPCPluginClientConfig struct {
+	ReconnectTimeout time.Duration
+	PingInterval     time.Duration
+}
+
+var defaultGRPCPluginClientConfig = GRPCPluginClientConfig{
+	ReconnectTimeout: time.Second * 20,
+	PingInterval:     time.Second * 2,
+}
+
+type GRPCPluginClientOption func(*GRPCPluginClientConfig)
+
+func WithReconnectConfig(reconnectTimeout time.Duration, pingInterval time.Duration) GRPCPluginClientOption {
+	return func(c *GRPCPluginClientConfig) {
+		c.ReconnectTimeout = reconnectTimeout
+		c.PingInterval = pingInterval
+	}
 }
 
 var _ grpc.ClientConnInterface = &GRPCPluginClient{}
 
-func newGRPCPluginClient(pc *plugin.Client, cc grpc.ClientConnInterface) (*GRPCPluginClient, error) {
+func newGRPCPluginClient(pc *plugin.Client, cc grpc.ClientConnInterface, options ...GRPCPluginClientOption) (*GRPCPluginClient, error) {
 	if pc == nil || cc == nil {
 		return nil, errors.New("plugin client or grpc client conn is nil")
 	}
 
+	config := defaultGRPCPluginClientConfig
+
+	for _, option := range options {
+		option(&config)
+	}
+
 	return &GRPCPluginClient{
-		pc: pc,
-		cc: cc,
+		pc:     pc,
+		cc:     cc,
+		config: config,
 	}, nil
 }
 
+func (g *GRPCPluginClient) waitForPluginToBeActive() error {
+	timeout := time.After(g.config.ReconnectTimeout)
+	for {
+		select {
+		case <-timeout:
+			return errors.New("plugin was not active in time")
+		default:
+			// we need to use a closure here to avoid race conditions
+			// as we cannot use defer in a for loop
+			shouldContinue, err := func() (bool, error) {
+				g.mu.RLock()
+				defer g.mu.RUnlock()
+				if g.pc == nil {
+					return true, nil
+				}
+
+				clientProtocol, err := g.pc.Client()
+				if err != nil {
+					return false, err
+				}
+
+				if err := clientProtocol.Ping(); err != nil {
+					time.Sleep(g.config.PingInterval)
+					return true, nil
+				}
+
+				return false, nil
+			}()
+
+			if err != nil {
+				return err
+			}
+
+			if !shouldContinue {
+				return nil
+			}
+		}
+	}
+}
+
+func (g *GRPCPluginClient) setClients(pc *plugin.Client, cc grpc.ClientConnInterface) {
+	// We need to lock here to avoid race conditions
+	// We potentially access the plugin clients during invokes
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	g.pc = pc
+	g.cc = cc
+}
+
 // Invoke implements grpc.ClientConnInterface.
 func (g *GRPCPluginClient) Invoke(ctx context.Context, method string, args any, reply any, opts ...grpc.CallOption) error {
-	if !g.isClosed.Load() {
+	if g.IsPluginProcessExited() {
+		if err := g.waitForPluginToBeActive(); err != nil {
+			return err
+		}
+	}
+
+	if g.isClosed.Load() {
 		return status.Error(codes.Unavailable, "plugin is not active")
 	}
 
+	g.mu.RLock()
+	defer g.mu.RUnlock()
 	return g.cc.Invoke(ctx, method, args, reply, opts...)
```

Comment:
I would suggest not returning the call result directly and perhaps wrapping the error, this is confusing on first read where defer is involved

Reply:
No in this case we shouldn't do anything with the error as the whole handling for this is done in the engine.

Reply:
Can we do this instead then? 
```go
g.mu.RLock()
defer g.mu.RUnlock()

if err := g.cc.Invoke(ctx, method, args, reply, opts...); err != nil {
    return err
}

return nil
```

Something spooky about using the mutex and defer in such a terse way, doesn't read clearly

Reply:
I prefer to leave it as it is.